internal/util/idlocker.go

```diff
@@ -59,3 +61,202 @@ func (vl *VolumeLocks) Release(volumeID string) {
 	defer vl.mux.Unlock()
 	vl.locks.Delete(volumeID)
 }
+
+type operation string
+
+var (
+	createOpt  operation = "create"
+	deleteOpt  operation = "delete"
+	cloneOpt   operation = "clone"
+	restoreOpt operation = "restore"
+	expandOpt  operation = "expand"
+)
+
+// OperationLock implements a map with atomic operations.
+type OperationLock struct {
+	locks map[operation]map[string]int
+	mux   sync.Mutex
+}
+
+// NewOperationLock returns new OperationLock.
+func NewOperationLock() *OperationLock {
+	lock := make(map[operation]map[string]int)
+	lock[createOpt] = make(map[string]int)
+	lock[deleteOpt] = make(map[string]int)
+	lock[cloneOpt] = make(map[string]int)
+	lock[restoreOpt] = make(map[string]int)
+	lock[expandOpt] = make(map[string]int)
+	return &OperationLock{
+		locks: lock,
+	}
+}
+
+// TryAcquire tries to acquire the lock for operating on volumeID and returns true if successful.
+// If another operation is already using volumeID, returns false.
+func (ol *OperationLock) TryAcquire(op operation, volumeID string) error {
+	ol.mux.Lock()
+	defer ol.mux.Unlock()
+	switch op {
+	case createOpt:
+		// check any delete operation is going on for given volume ID
```

Comment:
Instead of commenting what the next check obviously does, explain why this operation is not allowed to be queued. Here, and for each switch/case.